File: python/invariance/resources/agents.py
from __future__ import annotations

from typing import Any

from ..http_client import HttpClient
from ..types import AgentRecord, AgentMetrics, AgentActionTemplate, AgentActionPolicy


class AgentsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def create(self, opts: dict[str, str]) -> AgentRecord:
        return await self._http.post("/v1/agents", opts)

    async def list(self) -> list[AgentRecord]:
        return await self._http.get("/v1/agents")

    async def get(self, id: str) -> AgentRecord:
        return await self._http.get(f"/v1/agents/{id}")

    async def metrics(self) -> dict[str, Any]:
        return await self._http.get("/v1/agents/metrics")

    async def upsert_templates(
        self, agent_id: str, templates: list[AgentActionTemplate]
    ) -> dict[str, int]:
        return await self._http.put(
            f"/v1/agents/{agent_id}/templates", {"templates": templates}
        )

    async def get_templates(self, agent_id: str) -> list[AgentActionTemplate]:
        return await self._http.get(f"/v1/agents/{agent_id}/templates")

    async def upsert_policies(
        self, agent_id: str, policies: list[AgentActionPolicy]
    ) -> dict[str, int]:
        return await self._http.put(
            f"/v1/agents/{agent_id}/policies", {"policies": policies}
        )

    async def get_policies(self, agent_id: str) -> list[AgentActionPolicy]:
        return await self._http.get(f"/v1/agents/{agent_id}/policies")

File: packages/typescript/src/resources/agents.ts
import type { HttpClient } from '../http.js';
import type { AgentRecord, AgentMetrics, AgentActionTemplate, AgentActionPolicy } from '../types/agent.js';

export class AgentsResource {
  constructor(private http: HttpClient) {}

  async create(opts: {
    name: string;
    public_key?: string;
  }): Promise<AgentRecord> {
    return this.http.post<AgentRecord>('/v1/agents', opts);
  }

  async list(): Promise<AgentRecord[]> {
    return this.http.get<AgentRecord[]>('/v1/agents');
  }

  async get(id: string): Promise<AgentRecord> {
    return this.http.get<AgentRecord>(`/v1/agents/${id}`);
  }

  async metrics(): Promise<{ metrics: AgentMetrics[] }> {
    return this.http.get<{ metrics: AgentMetrics[] }>('/v1/agents/metrics');
  }

  async upsertTemplates(agentId: string, templates: AgentActionTemplate[]): Promise<{ updated: number }> {
    return this.http.put<{ updated: number }>(`/v1/agents/${agentId}/templates`, { templates });
  }

  async getTemplates(agentId: string): Promise<AgentActionTemplate[]> {
    return this.http.get<AgentActionTemplate[]>(`/v1/agents/${agentId}/templates`);
  }

  async upsertPolicies(agentId: string, policies: AgentActionPolicy[]): Promise<{ updated: number }> {
    return this.http.put<{ updated: number }>(`/v1/agents/${agentId}/policies`, { policies });
  }

  async getPolicies(agentId: string): Promise<AgentActionPolicy[]> {
    return this.http.get<AgentActionPolicy[]>(`/v1/agents/${agentId}/policies`);
  }
}
